snapshotter/demux/metrics/proxy.go

```diff
@@ -0,0 +1,137 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/containerd/containerd/log"
+)
+
+const (
+	// RequestPath is the path in the microVM hosting remote snapshotter metrics.
+	RequestPath = "http://localhost/metrics"
+	// MaxMetricsResponseSize is the limit in bytes for size of a metrics response from a remote snapshotter.
+	MaxMetricsResponseSize = 32768 // 32 KB
+)
+
+// Proxy represents a metrics proxy server for getting and serving remote snapshotter metrics.
+type Proxy struct {
+	// Client is the HTTP client used to send metrics requests to a remote snapshotter.
+	Client *http.Client
+	// Server is the proxy server on host for serving remote snapshotter metrics.
+	Server *http.Server
+	// Port is the port that the metrics proxy server listens on.
+	Port int
+	// Labels are labels used to apply to metrics.
+	Labels map[string]string
+}
+
+// NewProxy creates a new Proxy with initialized HTTP client and port.
+// dialer is used as the DialContext underlying the metrics HTTP client's RoundTripper.
+//
+// Reference: https://pkg.go.dev/net/http#Transport
+func NewProxy(portRange string, labels map[string]string, dialer func(context.Context, string, string) (net.Conn, error)) (*Proxy, error) {
+	port, err := findOpenPort(portRange)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Proxy{
+		Client: &http.Client{
+			Transport: &http.Transport{
+				DialContext: dialer,
+			},
+		},
+		Server: nil,
+		Port:   port,
+		Labels: labels,
+	}, nil
+}
+
+// findOpenPort uses net.Listen to determine if a port is available for hosting a metrics proxy server.
+func findOpenPort(portRange string) (int, error) {
+	ports := strings.Split(portRange, "-")
+	portRangeError := fmt.Errorf("invalid port range %s", portRange)
+	if len(ports) < 2 {
+		return 0, portRangeError
+	}
+	lower, err := strconv.Atoi(ports[0])
+	if err != nil {
+		return 0, portRangeError
+	}
+	upper, err := strconv.Atoi(ports[1])
+	if err != nil {
+		return 0, portRangeError
+	}
+
+	for port := lower; port <= upper; port++ {
+		ln, err := net.Listen("tcp", ":"+strconv.Itoa(port))
+		if err != nil {
+			continue
+		}
+		ln.Close()
+		if err != nil {
+			continue
+		}
+		return port, nil
+	}
```

Comment:
Can we assume that the TCP port in the range is "owned" by ourselves and manage them, instead of actually listening to see whether the port is available or not?

Reply:
Sure - so if we own this range, we are basically making the assumption that as the "owner" of the range, no other application will be listening on a port in this owned range outside of remote snapshotters metrics? this would then entail something tracking the state of used/open ports in that range, I imagine.

Reply:
updated in recent commit. follow up work required for where this owner should live and "freeing" of port once a proxy is shutdown https://github.com/firecracker-microvm/firecracker-containerd/issues/607